src/coreclr/jit/gentree.cpp

```diff
@@ -12685,63 +12685,132 @@ GenTree* Compiler::gtFoldExpr(GenTree* tree)
+        {
+            GenTree* thisOp = call->gtCallThisArg->GetNode();
+            GenTree* flagOp = call->gtCallArgs->GetNode();
+            GenTree* result = gtOptimizeEnumHasFlag(thisOp, flagOp);
 
-            // If either operand is known to be a RuntimeType, this can be folded
-            GenTree* result = gtFoldTypeEqualityCall(methodID, op1, op2);
             if (result != nullptr)
             {
                 return result;
             }
-            break;
         }
+    }
 
-        default:
+    // There was no intrinsic to fold, try the helper path.
+    // At present, we only fold the helpers for casts.
+    const CorInfoHelpFunc helper       = eeGetHelperNum(call->gtCallMethHnd);
+    bool                  overflowCast = false;
+    bool                  fromUnsigned = false;
+    var_types             castToType   = call->TypeGet();
+
+    assert(castToType == genActualType(castToType));
+
+    switch (helper)
+    {
+        case CORINFO_HELP_ULNG2DBL:
+            fromUnsigned = true;
             break;
+        case CORINFO_HELP_LNG2DBL:
+        case CORINFO_HELP_DBL2INT:
+        case CORINFO_HELP_DBL2LNG:
+            break;
+        case CORINFO_HELP_DBL2UINT:
+            castToType = TYP_UINT;
+            break;
+        case CORINFO_HELP_DBL2ULNG:
+            castToType = TYP_ULONG;
+            break;
+        case CORINFO_HELP_DBL2INT_OVF:
+        case CORINFO_HELP_DBL2LNG_OVF:
+            overflowCast = true;
+            break;
+        case CORINFO_HELP_DBL2UINT_OVF:
+            overflowCast = true;
+            castToType   = TYP_UINT;
+            break;
+        case CORINFO_HELP_DBL2ULNG_OVF:
+            overflowCast = true;
+            castToType   = TYP_ULONG;
+            break;
+        default:
+            // We cannot fold this call.
+            return call;
+    }
+
+    assert(call->fgArgInfo->ArgCount() == 1);
+    GenTree* arg = call->gtCallArgs->GetNode();
+
+    // We have a placeholder, find the real argument.
+    if (arg->OperIs(GT_ARGPLACE))
+    {
+        arg = call->fgArgInfo->GetArgNode(0);
     }
 
-    // Check for a new-style jit intrinsic.
-    const NamedIntrinsic ni = lookupNamedIntrinsic(call->gtCallMethHnd);
+    if (!arg->OperIsConst() || gtIsActiveCSE_Candidate(arg))
+    {
+        // We can only fold casts from constants.
+        // Also, we cannot fold if the argument can be CSEd.
+        return call;
+    }
 
-    if (ni == NI_System_Enum_HasFlag)
+    GenTree* cast = gtNewCastNode(call->TypeGet(), arg, fromUnsigned, castToType);
+    if (overflowCast)
     {
-        GenTree* thisOp = call->gtCallThisArg->GetNode();
-        GenTree* flagOp = call->gtCallArgs->GetNode();
-        GenTree* result = gtOptimizeEnumHasFlag(thisOp, flagOp);
+        cast->gtFlags |= GTF_OVERFLOW;
+    }
 
-        if (result != nullptr)
-        {
-            return result;
-        }
+    // Try and fold this cast.
+    GenTree* constVal = gtFoldExprConst(cast);
```

Comment:
Doing it this way leads to a somewhat confusing dump output where a "ghost node" appears to be folded.